### PR TITLE
msg/async/rdma: Introduce RDMAConnMgr + Debug prints

### PR DESCRIPTION
**Pull request by Adirl**

### CODE REVIEW

**Comment by Adirl** [comment]
@yuyuyu101 @amirv @orendu @saritz @DanielBar-On 

patchset number 4
Introduce RDMAConnTCP + Debug prints

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 1]
```diff
@@ -42,22 +48,31 @@ RDMAConnectedSocketImpl::RDMAConnectedSocketImpl(CephContext *cct, Infiniband* i
...
+}
+
+void RDMAConnectedSocketImpl::register_qp(QueuePair *qp)
+{
   notify_fd = dispatcher->register_qp(qp, this);
   dispatcher->perf_logger->inc(l_msgr_rdma_created_queue_pair);
   dispatcher->perf_logger->inc(l_msgr_rdma_active_queue_pair);
 }
 
+RDMAConnTCP::~RDMAConnTCP()
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: this should be RDMAConnTCP.cc

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 2]
```diff
@@ -88,7 +103,7 @@ void RDMAConnectedSocketImpl::get_wc(std::vector<ibv_wc> &w)
   w.swap(wc);
 }
 
-int RDMAConnectedSocketImpl::activate()
+int RDMAConnTCP::activate()
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nitto

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 3]
```diff
@@ -168,15 +183,15 @@ int RDMAConnectedSocketImpl::activate()
...
+    ldout(cct, 20) << __func__ << " handle fake send, wake it up. QP: " << *qp << dendl;
     submit(false);
   }
   active = true;
 
   return 0;
 }
 
-int RDMAConnectedSocketImpl::try_connect(const entity_addr_t& peer_addr, const SocketOptions &opts) {
+int RDMAConnTCP::try_connect(const entity_addr_t& peer_addr, const SocketOptions &opts) {
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nitto

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 4]
```diff
@@ -197,17 +212,17 @@ int RDMAConnectedSocketImpl::try_connect(const entity_addr_t& peer_addr, const S
...
     return r;
 
   worker->center.create_file_event(tcp_fd, EVENT_READABLE, con_handler);
   return 0;
 }
 
-void RDMAConnectedSocketImpl::handle_connection() {
-  ldout(cct, 20) << __func__ << " QP: " << my_msg.qpn << " tcp_fd: " << tcp_fd << " fd: " << notify_fd << dendl;
-  int r = infiniband->recv_msg(cct, tcp_fd, peer_msg);
+void RDMAConnTCP::handle_connection() {
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nitto

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 5]
```diff
@@ -575,7 +590,7 @@ void RDMAConnectedSocketImpl::fin() {
   }
 }
 
-void RDMAConnectedSocketImpl::cleanup() {
+void RDMAConnTCP::cleanup() {
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nitto

File: src/msg/async/rdma/RDMAConnectedSocketImpl.cc [hunk 6]
```diff
@@ -623,7 +638,7 @@ void RDMAConnectedSocketImpl::fault()
   notify();
 }
 
-void RDMAConnectedSocketImpl::set_accept_fd(int sd)
+void RDMAConnTCP::set_accept_fd(int sd)
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.cc]: nitto

File: src/msg/async/rdma/RDMAConnectedSocketImpl.h [hunk 7]
```diff
@@ -26,36 +26,39 @@
 class RDMAWorker;
 class RDMADispatcher;
 
+typedef Infiniband::QueuePair QueuePair;
+
 class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
+protected:
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.h]: one space before "protected"

File: src/msg/async/rdma/RDMAConnectedSocketImpl.h [hunk 8]
```diff
@@ -78,35 +81,19 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
...
+  virtual void fin();
+  virtual void cleanup() = 0;
+  virtual int try_connect(const entity_addr_t&, const SocketOptions &opt) = 0;
+  virtual ssize_t submit(bool more);
+  virtual int remote_qpn() = 0;
+
 };
 
 class RDMAServerSocketImpl : public ServerSocketImpl {
+protected:
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.h]: one space

File: src/msg/async/rdma/RDMAServerSocketImpl.cc [hunk 9]
```diff
@@ -34,7 +40,7 @@ RDMAServerSocketImpl::RDMAServerSocketImpl(CephContext *cct, Infiniband* i, RDMA
   ibdev->init(ibport);
 }
 
-int RDMAServerSocketImpl::listen(entity_addr_t &sa, const SocketOptions &opt)
+int RDMAServerConnTCP::listen(entity_addr_t &sa, const SocketOptions &opt)
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAServerSocketImpl.cc]: move to tcp .cc

File: src/msg/async/rdma/RDMAConnectedSocketImpl.h [hunk 10]
```diff
@@ -78,35 +81,19 @@ class RDMAConnectedSocketImpl : public ConnectedSocketImpl {
...
-    }
-    void close() {
-      active = false;
-    }
-  };
+  virtual void fin();
+  virtual void cleanup() = 0;
+  virtual int try_connect(const entity_addr_t&, const SocketOptions &opt) = 0;
+  virtual ssize_t submit(bool more);
+  virtual int remote_qpn() = 0;
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.h]: is this worth to define a virtual interface? I think we could have a variable in RDMAConnectedSocketImpl

File: src/msg/async/rdma/RDMAConnTCP.h [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+#include "common/debug.h"
+#include "common/errno.h"
+#include "msg/async/Stack.h"
+#include "Infiniband.h"
+#include "RDMAConnectedSocketImpl.h"
+
+class RDMAWorker;
+class RDMADispatcher;
+
+class RDMAConnTCP : public RDMAConnectedSocketImpl {
```
Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnTCP.h]: Maybe we define a RDMAConnMgr to encapsulate management info exchanges

Comment (yuyuyu101) [inline src/msg/async/rdma/RDMAConnTCP.h]: Inherit from RDMAConnectedSocketImpl makes it unclear. Because tcp or rdmacm is only used for management, right?

**Comment by amirv** [comment]
> this should be RDMAConnTCP.cc

I did it in purpose - to make the review easier (and rebase if needed).
If you don't mind, I prefer leaving this patch as is, and once it is merged I will send a cleanup commit to move code into the right files.

File: src/msg/async/rdma/RDMAConnectedSocketImpl.h (hunk 10 above)
Comment (amirv) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.h]: Correct - will change

File: src/msg/async/rdma/RDMAConnectedSocketImpl.h (hunk 8 above)
Comment (amirv) [inline src/msg/async/rdma/RDMAConnectedSocketImpl.h]: ack

File: src/msg/async/rdma/RDMAConnTCP.h (hunk 11 above)
Comment (amirv) [inline src/msg/async/rdma/RDMAConnTCP.h]: ack - adding RDMAConnMgr

**Comment by Adirl** [comment]
pushed new patchset and changed PR name
tested on vstart + traffic

**Comment by tchaikov** [comment]
@Adirl and @yuyuyu101 this breaks master, see https://jenkins.ceph.com/job/ceph-dev-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=centos7,DIST=centos7,MACHINE_SIZE=huge/6048//consoleFull

```
[ 38%] Building CXX object src/CMakeFiles/common-objs.dir/msg/async/net_handler.cc.o
In file included from /mnt/jenkins/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.0-1939-g4100803/rpm/el7/BUILD/ceph-12.0.0-1939-g4100803/src/msg/async/rdma/RDMAStack.h:30:0,
                 from /mnt/jenkins/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.0-1939-g4100803/rpm/el7/BUILD/ceph-12.0.0-1939-g4100803/src/msg/async/Stack.cc:22:
/mnt/jenkins/workspace/ceph-dev-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos7/DIST/centos7/MACHINE_SIZE/huge/release/12.0.0-1939-g4100803/rpm/el7/BUILD/ceph-12.0.0-1939-g4100803/src/msg/async/rdma/Infiniband.h:21:27: fatal error: rdma/rdma_cma.h: No such file or directory
 #include <rdma/rdma_cma.h>
                           ^
compilation terminated.
```

**Comment by tchaikov** [comment]
and the jenkins "make check" of this PR failed also, see https://jenkins.ceph.com/job/ceph-pull-requests/21036/

**Comment by tchaikov** [comment]
#14245 is posted to revert the offending commit.

**Comment by Adirl** [comment]
Checking and push new PR
Sorry

**Comment by amirv** [comment]
@yuyuyu101  ,
Sorry, an unused 'include "rdma_cma.h"' slipped into the patch.

We'll send a new PR soon...

@Adirl @orendu

**Pull request merged**